src/lib.rs

```diff
@@ -1,19 +1,521 @@
+//! # "#.trim());
+//! # /*
+//!
+//! */
+//! ```
+//!
+//! ### Verifying a Message
+//!
+//! ```rust
+//! use hex_literal::hex;
+//!
+//! use siwe::{SignInWithEthereum, Verify};
+//!
+//! use time::OffsetDateTime;
+//!
+//! # fn check_for_nonce_reuse(nonce: &str) -> anyhow::Result<String> { Ok(nonce.into()) }
+//!
+//! const MESSAGE: &str = r#"localhost wants you to sign in with your Ethereum account:
+//! 0x4b60ffAf6fD681AbcC270Faf4472011A4A14724C
+//!
+//! Allow localhost to access your orbit using their temporary session key: did:key:z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg#z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg
+//!
+//! URI: did:key:z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg#z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg
+//! Version: 1
+//! Chain ID: 1
+//! Nonce: PPrtjztx2lYqWbqNs
+//! Issued At: 2021-12-20T12:29:25.907Z
+//! Expiration Time: 2021-12-20T12:44:25.906Z
+//! Resources:
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#put
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#del
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#get
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#list"#;
+//!
+//! const SIGNATURE: [u8; 65] = hex!("20c0da863b3dbfbb2acc0fb3b9ec6daefa38f3f20c997c283c4818ebeca96878787f84fccc25c4087ccb31ebd782ae1d2f74be076a49c0a8604419e41507e9381c");
+//!
+//! # fn run() -> anyhow::Result<siwe::Message> {
+//! # Ok({
+//! // `SignInWithEthereum` is the main entry point for both building and validation.
+//! let verifier = SignInWithEthereum::default()
+//!
+//!     // Common options for both building and validating messages:
+//!     .domain("localhost")?
+//!     .uri("did:key:z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg#z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg")?
+//!     .chain_id(1)
+//!
+//!     // Indicate that we want to verify a message (not build one.)
+//!     .verifier(MESSAGE)?
+//!
+//!     // Optionally pick a custom time, or leave it as the current time.
+//!     .valid_at(OffsetDateTime::from_unix_timestamp(1640003425)?)
+//!
+//!     // Customize which fields to verify (or leave the defaults.)
+//!     .statement(Verify::Never);
+//!
+//! // It is important to check that this message's nonce has never been used.
+//! let nonce = check_for_nonce_reuse(verifier.nonce())?;
+//!
+//! verifier
+//!     .verify(&nonce, &SIGNATURE)?
+//! # })
+//! # }
+//! # run().unwrap();
+//! ```
 use core::{
     convert::Infallible,
     fmt::{self, Display, Formatter},
     str::FromStr,
 };
 use hex::FromHex;
-use http::uri::{Authority, InvalidUri};
+use http::{
+    uri::{Authority, InvalidUri},
+    Uri,
+};
 use iri_string::types::UriString;
+use std::convert::TryFrom;
 use thiserror::Error;
+use time::Duration;
 use time::OffsetDateTime;
 
 pub mod nonce;
 pub mod rfc3339;
 
 pub use rfc3339::TimeStamp;
 
+#[derive(Debug, Clone, Copy)]
+enum Time {
```

Comment:
We should only manipulate timestamps

Reply:
The `Time` enum is mostly to allow downstream crates to specify either an offset from now (eg. expire in 4 hours) or an exact time.

Are you suggesting I store a `TimeStamp` inside `Time`, that we shouldn't allow offsets, or something else?

Reply:
I don't think we should allow offsets. It's unclear as to what the offset is from. Is it from the moment the method to add the offset is called? Is it at message build time? Is it at some other time? Realistically I think we should encourage users to know precisely what timestamp they are setting, otherwise it could lead to unexpected behaviour.

Reply:
Documenting the behavior would clear that up, but I do understand your concern. I believe the library should make the common use cases easy (like "expire in X minutes"), but if the verdict is to only support explicit timestamps, that's what I'll do.

Reply:
While I agree you could make it clear in documentation, not everyone reads it or even finds it easy to read documentation. Do you think `expire in X minutes` is hard at the moment?
```rust
Message {
  ...
  expiration_time: OffsetDateTime::now_utc() - Duration::from_secs(60*10),
}
```
I would prefer to keep it explicit 
like this, since it's obvious exactly when the SIWE message is expiring, and it's already quite succinct thanks to the `time` crate.